### DISCUSSION TITLE
Iterate through multiple BMSes

### DISCUSSION THREAD

**Discussion by dan-cristian** [syssi/esphome-jbd-bms#32] [Ideas]
I have many battery packs/BMSes and I don't really need a very frequent read on BMS values.
I would like to use one single ESP32 board to iterate through multiple BMS boards so I reduce the number of ESP boards I currently use (9 ESPs). I only need the MQTT topic prefix to change when connecting to a BMS board, I currently use board MAC as prefix.

I logged the idea also in JK repository.
https://github.com/syssi/esphome-jk-bms/discussions/281

**Comment by syssi** [comment]
Do you use a BLE or UART connection?

**Reply by dan-cristian** [reply]
I use BLE for all of them.

**Reply by syssi** [reply]
How many JBD and how many BMS devices?

**Reply by dan-cristian** [reply]
I have 6 with JBD and 2 with JK  (and one more with JK in my plans).

**Reply by dan-cristian** [reply]
![image](https://user-images.githubusercontent.com/10730231/224996837-fcce1fe9-6f2a-49f8-a25e-eaf7ed61af97.png)

**Reply by syssi** [reply]
The official BLE client component refuses to setup more then 3 BLE client connections. I've removed this limitation because it's about simultaneous connections if we are lucky. Could you try this YAML:

https://github.com/syssi/esphome-jbd-bms/blob/test-many-ble-connections/esp32-ble-example-multiple-devices.yaml

Please use the switches to enable one BLE connection after the other. Are you able to talk to each device step by step? If this is working we could implement a iterator to cycle through all devices while using a single BLE client connection.

**Reply by dan-cristian** [reply]
I was thinking to use more a step by step approach / iterator rather than simultaneous connections. I'll try the yaml also.

**Reply by syssi** [reply]
This is the step by step / iterator approach. All BLE client connection switches are disabled per default. Please update the MAC addresses to the one of your devices and flash the YAML. The next step is testing the iterator approach manually:

1. Enable the BLE connection switch of BMS0
2. Wait 20 seconds
3. Disable the BLE connection switch of BMS0
4. Enable the BLE connection switch of BMS1
5. Wait 20 seconds
6. Disable the BLE connection switch of BMS1
...

**Reply by syssi** [reply]
Does it work? :-)

**Comment by dan-cristian** [comment]
Have not tried yet, need to finish building my last big battery pack (6kw)
and will try after that. Next weekend most likely.

**Comment by walcase** [comment]
I also like the idea of iterating through BMSs using a single ESP32. Can it be done? Is there an example yaml file I can try? 

Thanks syssl for this software. It works great!!!

**Reply by syssi** [reply]
Please try this example: https://github.com/syssi/esphome-jbd-bms/pull/33

**Reply by walcase** [reply]
I just noticed that the changes were committed to "main" If this is the case ignore my previous messages on a new thread.

 So I am setup for 3 BMS's via bluetooth. This works.  When I add a forth the YAML code validator complains that only 3 BLE connections are supported. Maybe I can ignore this and install the code anyway?

WC

**Comment by walcase** [comment]
Using the example yaml, can't seem to find the github sources referenced by "test-many-ble-connections".."main" works find.

**_couldn't find remote ref test-many-ble-connections.
    source: github://syssi/esphome-jbd-bms@test-many-ble-connections_**

Not sure if I am missing something??

WC

**Reply by syssi** [reply]
Ups. It looks like I've deleted the feature branch with the patched BLE component. Wait a sec. I will prepare a new one.

**Reply by syssi** [reply]
I've updated and pushed the feature branch.

https://github.com/syssi/esphome-jbd-bms/pull/79
https://github.com/syssi/esphome-jbd-bms/tree/test-many-ble-connections

Please give it another try using this example:

https://github.com/syssi/esphome-jbd-bms/blob/test-many-ble-connections/esp32-ble-example-multiple-devices.yaml

**Reply by walcase** [reply]
I have 4 BMS's working now using the branch..Update times vary between 15sec and 2min according to Home Assistant. Will add more over the next few days..Would like to get 16 BMS's on the one esp32. Will add more over the next week or so.

Thanks!!
WC

**Reply by walcase** [reply]
I have 5 BMSs working now but I have two questions;

1. Is the branch connecting and disconnecting the BLE as it loops through the list of BMSs? I as just wondering as I thought it would take longer for updates, currently the updates can be as short as 10 sec.
2. What would  be the effect, if any,  of changing the "update interval" to a larger value?
                        _jbd_bms_ble:
                       - id: bms0
                       ble_client_id: client0
                       **update_interval: 2s**_

These questions come from wondering just how many BMSs can one ESP32 do..

Thanks!!
WC

**Reply by syssi** [reply]
1. No. The YAML is just a proof of concept. You have to enable the BLE client connections manually at the moment. If it's working in general you can use a script at the bottom of the YAML.
2. The BMS gets polled. If you change the `update_interval` you are changing the polling frequency.

**Reply by walcase** [reply]
I will see how many BMS's will work with simultaneous connections until the memory runs out. Then figure out if I even need a script to loop through them one at a time instead.

**Reply by syssi** [reply]
The ESP32 supports 3 simultaneous BLE client connections. Please use the YAML to proof we can talk to more devices if we don't exceed this limit.

**Reply by syssi** [reply]
See https://github.com/esphome/esphome/pull/7088

**Comment by walcase** [comment]
In Home Assistant i did some manual testing and when the "enable bluetooth connection swiitch" is set to off the sensors for that BMS change to "Unknown". I assume this would be the same if it happened in the YAML script through iteration.

On the upside I have had **5 simultaneous** BT connections running solid for the better part of 48hrs. My understanding is that the 3 connection limit is not a limit of hardware or software(ESP32)  but defined to conserve memory. In my case I run an ESP32-S3 with only the very basic services and is only used for this specific application.
  
If I can get 6 simultaneous BT connections then that would work for me. The key was removing the 3 connection limit.
I  will continue to update this thread as I add more BMS's.

**Reply by syssi** [reply]
Interesting. :-) I thought the limit came from here and is a hard limit: https://docs.espressif.com/projects/esp-faq/en/latest/software-framework/ble-bt.html#does-esp32-support-multiple-clients-connecting-at-the-same-time-when-used-as-a-ble-server-how-to-realize-it

**Reply by walcase** [reply]
In that same FAQ #21;
https://docs.espressif.com/projects/esp-faq/en/latest/software-framework/ble-bt.html#how-many-bluetooth-clients-can-be-connected-to-esp32

I also found this reference which seems to lineup with the FAQ;
https://linuxhint.com/esp32-maximum-bluetooth-connections/

It would be nice to have the option to add more than three connections in the main branch noting that anything over 3 may be unstable depending on your application. I have 5 working now and will add a sixth today.

**Reply by syssi** [reply]
Good catch but please keep in mind we are talking about BLE client connections. Some answers are about running the BLE stack "in server mode".

**Reply by walcase** [reply]
Just added a sixth BMS and started to get connection failures. Moved back to 5 and all seems solid again. So in my setup 5 appears to be the limit. I can work with that. Good thing ESP32's are cheap!!

**Reply by walcase** [reply]
syssl, Any thought about having the option to add more than three BT connections in the main branch?  A bit worried the test branch will disappear some day. I have had 5 simultaneous BT connections working for weeks now with no issues.

Thanks again for your work on this!!

**Reply by syssi** [reply]
See https://github.com/esphome/esphome/pull/7088

**Comment by ExR90** [comment]
@syssi 

Any plans to make this part of the main branch with normal features? I tested this branch with 4 BMS and it works, but it doesn't have all the sensors the main branch does, like the per-cell-details etc. I am using 4 ESP32 to run things now, would be nice to only use 1. I don't care about super fast polling, a couple times a minute is more than enough if that matters.

**Reply by syssi** [reply]
I don't want to maintain a fork of the official `ble_client` component here. The restriction must be relaxed upstream (at the ESPHome core). Just dropping the restriction won't be accepted most likely. A small improvement needs to be implemented to count the active connections and warn the user if the limited is exceeded.

**Reply by walcase** [reply]
I would second any push to update the ble_client dropping this restriction. I have many essp32s3 working with 4 and 5 connections/BMSs without issue.

**Reply by ExR90** [reply]
> I don't want to maintain a fork of the official `ble_client` component here. The restriction must be relaxed upstream (at the ESPHome core). Just dropping the restriction won't be accepted most likely. A small improvement needs to be implemented to count the active connections and warn the user if the limited is exceeded.

Is there a way to at least get all of the sensors as the main branch in this test branch?

**Reply by syssi** [reply]
They are available. You can simply add these sensors to the YAML.

**Comment by walcase** [comment]
Looks like the branch has been deleted?

_INFO ESPHome 2025.5.2
INFO Reading configuration /config/esphome/esphome-web-5eb7c8.yaml...
INFO Updating https://github.com/syssi/esphome-jbd-bms.git@test-many-ble-connections
Failed config

external_components: [source /config/esphome/esphome-web-5eb7c8.yaml:53]
  - 
    couldn't find remote ref test-many-ble-connections.
    source: github://syssi/esphome-jbd-bms@test-many-ble-connections
    refresh: 0s_

Thanks WC

**Comment by walcase** [comment]
Migrated to main branch from "test-many-ble-connections" since simultaneous BT connections limit of three(3) removed from main.  Managing sixteen(16) BMS's with ESP32s via simultaneous BT connections, four(4) connections per esp32. Working great!!

**Reply by syssi** [reply]
Nice!
